File: src/main/kotlin/school/charset/app/domain/exercise/ExerciseAttempt.kt
package school.charset.app.domain.exercise

import school.charset.app.domain.encoding.CodePoint
import school.charset.app.domain.encoding.Encoding
import kotlin.time.Instant

data class ExerciseAttempt(
    val id: Long,
    val userId: Long,
    val module: ExerciseModule,
    val level: Int,
    val codePoint: CodePoint,
    val encoding: Encoding,
    val correct: Boolean,
    val finalized: Boolean,
    val durationMs: Int?,
    val steps: List<AttemptStep>,
    val createdAt: Instant,
)

File: lib/charset/schema/exercise_attempt.ex
defmodule Charset.Schema.ExerciseAttempt do
  @moduledoc "Ecto schema for the `exercise_attempts` parent table."

  use Ecto.Schema

  schema "exercise_attempts" do
    field :token, :string
    field :module_id, :string
    field :level, :integer
    field :code_point, :integer
    field :encoding, :string
    field :correct, :boolean, default: false
    field :finalized, :boolean, default: false
    field :duration_ms, :integer

    has_many :steps, Charset.Schema.AttemptStep, foreign_key: :attempt_id

    timestamps(updated_at: false, type: :utc_datetime)
  end
end
